### PR TITLE
CAL-134: Mapped CSEXRA, CSDIDA, PIAIMC, and HISTOA TREs to ext.nitf.* fields

### PR DESCRIPTION
**Pull request by troymohl**
#### What does this PR do?
Adds mapping for non-taxonomy NITF TRE fields into ext.nitf fields on the metacard.

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@glenhein 
@bdeining
@peterhuffer  

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@kcwire
@millerw8

#### How should this be tested?
Run full build including JUnit and i-tests.

#### What are the relevant tickets?
https://codice.atlassian.net/browse/CAL-134

#### Checklist:
- [X] Documentation Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/CsdidaAttribute.java [hunk 1]
```diff
@@ -33,31 +35,120 @@
...
+
+    static final CsdidaAttribute PLATFORM_ID = new CsdidaAttribute(Isr.PLATFORM_ID,
             "PLATFORM_CODE_VEHICLE_ID",
             CsdidaAttribute::getPlatformIdFunction,
             new IsrAttributes().getAttributeDescriptor(Isr.PLATFORM_ID),
             "platformCodeVehicleId",
-            "csdida.");
+            ATTRIBUTE_NAME_PREFIX);
+
+    static final CsdidaAttribute DAY = new CsdidaAttribute("day-dataset-collection",
```
Comment (peterhuffer) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/CsdidaAttribute.java]: Typically we have been following camel case convention for these NITF field names. For example, in NitfHeaderAttribute, FileProfileName is added as `ext.nitf.fileProfileName`.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/CsdidaAttribute.java [hunk 2]
```diff
@@ -33,31 +35,120 @@
...
+            BasicTypes.STRING_TYPE,
+            ATTRIBUTE_NAME_PREFIX);
+
+    static final CsdidaAttribute PRODUCT_ID = new CsdidaAttribute("product-id",
+            PRODUCT_ID_NAME,
+            tre -> TreUtility.parseString(tre, PRODUCT_ID_NAME),
+            BasicTypes.STRING_TYPE,
+            ATTRIBUTE_NAME_PREFIX);
+
+    static final CsdidaAttribute TIME = new CsdidaAttribute("iamge-start-time",
```
Comment (peterhuffer) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/CsdidaAttribute.java]: typo: iamge-start-time

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/TreUtility.java [hunk 3]
```diff
@@ -55,4 +61,61 @@ public static Serializable getTreValue(TreGroup tre, String key) {
...
+                return sdf.parse(dateStr);
+            } catch (ParseException e) {
+                LOGGER.debug("Unable to parse date {} according to format {}",
+                        dateStr,
+                        TRE_DATE_FORMAT,
+                        e);
+                return null;
+            }
+        } else {
+            return null;
```
Comment (glenhein) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/TreUtility.java]: optional: remove the two 'return null' statements and the else-block, replace with a single 'return null' and the end of the method.

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/common/CsexraAttributeTest.java [hunk 4]
```diff
@@ -229,12 +229,873 @@ public void testSnowDepthMaxCategoryOther() throws NitfFormatException {
...
+    }
+
+    @Test
+    public void testGeoMeanVertGSDNA() throws NitfFormatException {
+        when(tre.getFieldValue(CsexraAttribute.GEO_MEAN_VERT_GSD_NAME)).thenReturn("N/A");
+
+        Serializable actual = CsexraAttribute.GEO_MEAN_VERT_GSD.getAccessorFunction()
+                .apply(tre);
+
+        assertThat(actual, is(instanceOf(String.class)));
```
Comment (glenhein) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/common/CsexraAttributeTest.java]: do you really need the instanceOf check? seems redundant

Comment (troymohl) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/common/CsexraAttributeTest.java]: Should I remove all of the instanceOf tests? Those were part of the test class, so I kept with the existing pattern.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/TreUtility.java (hunk 3 above)
Comment (troymohl) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/TreUtility.java]: Fixed.

**Comment by troymohl** [comment]
@peterhuffer good catch on adding those descriptors. Fixed.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/CsdidaAttribute.java (hunk 2 above)
Comment (troymohl) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/CsdidaAttribute.java]: fixed.

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/common/CsexraAttributeTest.java (hunk 4 above)
Comment (glenhein) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/common/CsexraAttributeTest.java]: it's a judgement call, i'll leave it up to you

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/CsexraAttribute.java [hunk 5]
```diff
@@ -85,20 +129,159 @@
...
                     "snowDepthMin",
                     ATTRIBUTE_NAME_PREFIX);
 
     static final CsexraAttribute SNOW_DEPTH_MAX =
             new CsexraAttribute(Isr.SNOW_DEPTH_MAX_CENTIMETERS,
-                    SNOW_DEPTH_CAT,
+                    SNOW_DEPTH_CAT_NAME,
                     getSnowDepthAccessorFunction(Pair::getRight),
                     new IsrAttributes().getAttributeDescriptor(Isr.SNOW_DEPTH_MAX_CENTIMETERS),
                     "snowDepthMax",
```
Comment (peterhuffer) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/CsexraAttribute.java]: snow-depth-max

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/HistoaAttribute.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    static final HistoaAttribute REMAP_FLAG = new HistoaAttribute("system-specific-remap",
+            REMAP_FLAG_NAME,
+            tre -> TreUtility.parseString(tre, REMAP_FLAG_NAME),
+            BasicTypes.STRING_TYPE,
+            ATTRIBUTE_NAME_PREFIX);
+
+    static final HistoaAttribute LUTID = new HistoaAttribute("data-mapping-id",
+            LUTID_NAME,
+            tre -> TreUtility.parseInteger(tre, LUTID_NAME),
+            BasicTypes.STRING_TYPE,
```
Comment (roelens8) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/HistoaAttribute.java]: I think this should be BasicTypes.INTEGER_TYPE

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/CsexraAttribute.java (hunk 5 above)
Comment (troymohl) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/CsexraAttribute.java]: Fixed.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/CsexraAttribute.java [hunk 7]
```diff
@@ -85,20 +129,159 @@
 
     static final CsexraAttribute SNOW_DEPTH_MIN =
             new CsexraAttribute(Isr.SNOW_DEPTH_MIN_CENTIMETERS,
-                    SNOW_DEPTH_CAT,
+                    SNOW_DEPTH_CAT_NAME,
                     getSnowDepthAccessorFunction(Pair::getLeft),
                     new IsrAttributes().getAttributeDescriptor(Isr.SNOW_DEPTH_MIN_CENTIMETERS),
                     "snowDepthMin",
```
Comment (troymohl) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/CsexraAttribute.java]: Fixed.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/ImageAttribute.java [hunk 8]
```diff
@@ -211,92 +211,92 @@
...
-            "numberOfImageComments",
+            "number-of-image-comments",
             "NICOM",
             segment -> segment.getImageComments()
                     .size(),
             BasicTypes.INTEGER_TYPE,
             ATTRIBUTE_NAME_PREFIX);
 
-    public static final ImageAttribute IMAGE_COMMENT_1 = new ImageAttribute("imageComment1",
+    public static final ImageAttribute IMAGE_COMMENT_1 = new ImageAttribute("image-comment1",
```
Comment (peterhuffer) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/ImageAttribute.java]: we did `image-identifier-2` and `image-identifier-1`, do we want to make these `image-comment-1`?

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AcftbAttribute.java [hunk 9]
```diff
@@ -42,73 +42,73 @@
             "AC_MSN_ID",
             tre -> TreUtility.getTreValue(tre, "AC_MSN_ID"),
             new IsrAttributes().getAttributeDescriptor(Isr.MISSION_ID),
-            "aircraftMissionId",
+            "aircraft-mission-id",
```
Comment (kcwire) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AcftbAttribute.java]: Why are we changing ACFTB TREs in this ticket?

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java [hunk 10]
```diff
@@ -37,45 +37,45 @@
     /*
      * Non-normalized attributes
      */
-    public static final AimidbAttribute ACQUISITION_DATE = new AimidbAttribute("acquisitionDate",
+    public static final AimidbAttribute ACQUISITION_DATE = new AimidbAttribute("acquisition-date",
```
Comment (kcwire) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java]: Same question here?

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/CsexraAttribute.java [hunk 11]
```diff
@@ -31,27 +31,71 @@
...
 /**
  * TRE for "Exploitation Reference Data"
  */
-class CsexraAttribute extends NitfAttributeImpl<Tre> {
+public class CsexraAttribute extends NitfAttributeImpl<Tre> {
 
     private static final List<NitfAttribute<Tre>> ATTRIBUTES = new LinkedList<>();
 
-    static final String SNOW_DEPTH_CAT = "SNOW_DEPTH_CAT";
+    static final String SNOW_DEPTH_CAT_NAME = "SNOW_DEPTH_CAT";
```
Comment (kcwire) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/CsexraAttribute.java]: is it intentional all of these constants are package-private? How would an integrator get the attribute for SNOW_DEPTH_CAT?

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/HistoaAttribute.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+import ddf.catalog.data.impl.BasicTypes;
+
+/**
+ * TRE for "Softcopy History Tagged Record Extension"
+ */
+public class HistoaAttribute extends NitfAttributeImpl<Tre> {
+
+    private static final List<NitfAttribute<Tre>> ATTRIBUTES = new LinkedList<>();
+
+    static final String SYSTYPE_NAME = "SYSTYPE";
```
Comment (kcwire) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/HistoaAttribute.java]: more package-private?

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderAttribute.java [hunk 13]
```diff
@@ -56,7 +56,7 @@
                     header -> header.getFileType()
                             .name(),
                     new MediaAttributes().getAttributeDescriptor(Media.FORMAT),
-                    "fileProfileName",
+                    "file-profile-name",
```
Comment (kcwire) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderAttribute.java]: More changes that appear unrelated to the TREs being added.

**Comment by troymohl** [comment]
This PR will be closed once the individual PRs have been opened to address @kcwire request.

**Comment by troymohl** [comment]
This PR has been broken up and submitted in the following Pull Requests:
https://github.com/codice/alliance/pull/179
https://github.com/codice/alliance/pull/180
https://github.com/codice/alliance/pull/181
https://github.com/codice/alliance/pull/182
https://github.com/codice/alliance/pull/185

**Pull request closed without merging**